dot2ubi.1
.TH DOT2UBI 1 "2014-04-23" "v1.0" ""
.nh
.SH "NAME"
dot2ubi \- Render Graphviz files in 3-D using Ubigraph
.SH "SYNOPSIS"
dot2ubi [\fIfilename\fR]
.SH "DESCRIPTION"
\fBdot2ubi\fR plots graphs (as in ``edges and vertices'', not ``pie charts
and bar charts'') with a 3-D layout.  It reads a graph description in
Graphviz (``dot'') format and communicates the graph to a Ubigraph
server application for rendering.
.SH "OPTIONS"
There are no command-line options.  \fBdot2ubi\fR accepts the name of a
``dot'' file.  If not provided, the program reads from the standard
input device.
.SH "NOTES"
Graphviz and Ubigraph support different features.  \fBdot2ubi\fR tries
hard to map node and vertex attributes from what makes sense to
Graphviz into what makes sense to Ubigraph.  For example, planar node
shapes such as \f(CW\*(C`box\*(C'\fR and \f(CW\*(C`circle\*(C'\fR are mapped to solid node shapes
such as \f(CW\*(C`cube\*(C'\fR and \f(CW\*(C`sphere\*(C'\fR.
.PP
Explicitly specified colors are honored, but the default colors are
different due to Graphviz rendering atop a white background and
Ubigraph rendering atop a black background.  Also, Graphviz nodes
include both border and fill colors while Ubigraph nodes contain only
fill colors.  \fBdot2ubi\fR draws both nodes and edges in white unless
otherwise specified.
.PP
Both directed and undirected graphs are supported.
.PP
All edge stroke styles are supported.
.PP
Arrow sizes are supported.  Arrow shapes are ignored.
.PP
Ubigraph does not support double-ended arrows (\f(CW\*(C`dir=both\*(C'\fR in
Graphviz); these are rendered as single-ended arrows.
.PP
Ubigraph does not support Graphviz's subgraphs or clusters.  Nodes
within subgraphs/clusters are drawn.  Only their grouping is ignored.
.PP
URLs can be associated with nodes via either the
\f(CW\*(C`href\*(C'\fR or the \f(CW\*(C`URL\*(C'\fR attribute.
.SH "AUTHOR"
Scott Pakin, \fIscott+dt2ub@pakin.org\fR
.SH "COPYRIGHT AND LICENSE"
Copyright 2014 Scott Pakin
.PP
This software is released under a \s-1BSD\s0
license <http://opensource.org/licenses/BSD-2-Clause>.
.SH "SEE ALSO"
\fIdot\fR\|(1),
the Graphviz home page <http://www.graphviz.org/>,
the Ubigraph home page <http://ubietylab.net/ubigraph/>


dot2ubi.c
/*******************************************************
 * Visualize a Graphviz dot file in 3-D using Ubigraph *
 * By Scott Pakin <scott+dt2ub@pakin.org>              *
 *******************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <graphviz/cgraph.h>
#include <graphviz/color.h>
#include <UbigraphAPI.h>

/* Produce a Ubigraph ID from a Graphviz object. */
#define UBID(N) ((int)AGID(N))

/* Get the value of a node/edge attribute and return true if not NULL or empty. */
#define GET_NODE_VALUE(KEY) ((value = agget(node, KEY)) != NULL && value[0] != '\0')
#define GET_EDGE_VALUE(KEY) ((value = agget(edge, KEY)) != NULL && value[0] != '\0')

extern int colorxlate(const char *str, gvcolor_t *color, color_type_t target_type);
extern void setColorScheme (char *s);

const char *progname;   /* Name of this program */

/* Associate planar with solid shapes. */
typedef struct {
  const char *planar;      /* Name of a planar shape */
  const char *solid;       /* Name of a solid shape */
} planar_solid_t;

/* Define a mapping from planar to solid shapes. */
planar_solid_t **planar_to_solid;
size_t planar_to_solid_len;

/* Initialize the planar-to-solid mapping. */
void initialize_planar_to_solid (void)
{
  static planar_solid_t initial_map[] = {
    { "box",            "cube"         },
    { "polygon",        "dodecahedron" },
    { "ellipse",        "sphere"       },
    { "circle",         "sphere"       },
    { "point",          "sphere"       },
    { "egg",            "sphere"       },
    { "triangle",       "octahedron"   },
    { "plaintext",      "cube"         },
    { "diamond",        "octahedron"   },
    { "trapezium",      "cube"         },
    { "parallelogram",  "cube"         },
    { "house",          "octahedron"   },
    { "octagon",        "icosahedron"  },
    { "doublecircle",   "sphere"       },
    { "doubleoctagon",  "icosahedron"  },
    { "tripleoctagon",  "icosahedron"  },
    { "invtriangle",    "octahedron"   },
    { "invtrapezium",   "cube"         },
    { "invhouse",       "octahedron"   },
    { "Mdiamond",       "octahedron"   },
    { "Msquare",        "cube"         },
    { "Mcircle",        "sphere"       },
    { "none",           "cube"         },
    { "record",         "cube"         },
    { "Mrecord",        "cube"         }
  };
  size_t i;

  planar_to_solid_len = sizeof(initial_map)/sizeof(planar_solid_t);
  planar_to_solid = (planar_solid_t **) malloc(planar_to_solid_len*sizeof(planar_solid_t *));
  if (planar_to_solid == NULL) {
    perror("malloc");
    exit(1);
  }
  for (i = 0; i < planar_to_solid_len; i++)
    planar_to_solid[i] = &initial_map[i];
}

/* Read a dot file into memory and a return the associated graph.
 * Abort on error. */
Agraph_t *read_dot (const char *dotfilename)
{
  int use_stdin;    /* 1=read from stdin; 0=read from a file */
  FILE *dotfile;
  Agraph_t *graph;

  /* Open the input file.  "-" means standard input. */
  use_stdin = dotfilename[0] == '-' && dotfilename[1] == '\0';
  if (use_stdin)
    dotfile = stdin;
  else {
    dotfile = fopen(dotfilename, "r");
    if (!dotfile) {
      perror("fopen");
      exit(1);
    }
  }

  /* Parse the input file. */
  graph = agread(dotfile, &AgDefaultDisc);
  if (!graph) {
    fprintf(stderr, "%s: Failed to parse %s\n", progname, use_stdin ? "standard input" : dotfilename);
    exit(1);
  }

  /* Close the input file. */
  if (!use_stdin)
    fclose(dotfile);
  return graph;
}

/* Initialize the Ubigraph canvas. */
void initialize_ubigraph (Agraph_t *graph)
{
  /* Set defaults that resemble dot's defaults.  There are some
   * differences due to what makes sense in 3-D vs. 2-D and what makes
   * sense on a black background vs. a white background. */
  ubigraph_clear();
  ubigraph_set_vertex_style_attribute(0, "color", "#ffffff");  /* dot: no fill but white background */
  ubigraph_set_vertex_style_attribute(0, "shape", "sphere");   /* dot: ellipse */
  ubigraph_set_vertex_style_attribute(0, "size", "0.75");   /* dot: 0.75 by 0.5 */
  ubigraph_set_vertex_style_attribute(0, "fontfamily", "Times-Roman");   /* Same as dot */
  ubigraph_set_vertex_style_attribute(0, "fontcolor", "#ffffff");   /* dot: black */
  ubigraph_set_vertex_style_attribute(0, "fontsize", "14");   /* Same as dot */
  if (agisdirected(graph)) {
    ubigraph_set_edge_style_attribute(0, "arrow", "true");   /* Same as dot */
    ubigraph_set_edge_style_attribute(0, "arrow_position", "1.0");   /* Same as dot */
  }
  ubigraph_set_edge_style_attribute(0, "color", "#ffffff");  /* dot: black */
  ubigraph_set_edge_style_attribute(0, "fontfamily", "Times-Roman");   /* Same as dot */
  ubigraph_set_edge_style_attribute(0, "fontcolor", "#ffffff");   /* dot: black */
  ubigraph_set_edge_style_attribute(0, "fontsize", "14");   /* Same as dot */
  initialize_planar_to_solid();
}

/* Convert a color to #RRGGBB format.  Return a pointer to a static string. */
char *convert_color (const char *color_name)
{
  static char rgb[10];
  gvcolor_t color;

  if (colorxlate(color_name, &color, RGBA_BYTE) != COLOR_OK) {
    fprintf(stderr, "%s: Failed to identify color \"%s\"\n", progname, color_name);
    exit(1);
  }
  sprintf(rgb, "#%02X%02X%02X",
          color.u.rgba[0],
          color.u.rgba[1],
          color.u.rgba[2]);
  return rgb;
}

/* Convert a planar shape to a solid shape. */
const char *convert_shape (const char *planar)
{
  size_t i;

  for (i = 0; i < planar_to_solid_len; i++)
    if (!strcmp(planar_to_solid[i]->planar, planar)) {
      /* Found -- bubble up the association for faster lookup next time. */
      const char *solid = planar_to_solid[i]->solid;
      if (0 && i > 0) {
        planar_solid_t *prev = planar_to_solid[i - 1];
        planar_to_solid[i - 1] = planar_to_solid[i];
        planar_to_solid[i] = prev;
      }
      return solid;
    }
  return "torus";   /* Use "torus" as the not-found solid. */
}

/* Replace "\N" in a string with the node name and return a static string. */
char *replace_N (char *str, char *name)
{
  static char *result = NULL;    /* Resulting string */
  static size_t allocated = 0;   /* Bytes allocated for the above */
  size_t namelen;                /* Length of replacement string */
  char *found;                   /* Pointer to first "\N" in the input string */
  char *resultp;                 /* Pointer into result */

  /* Common case -- no "\N" in the input string. */
  found = strstr(str, "\\N");
  if (!found)
    return str;

  /* Replace each occurrence in turn. */
  namelen = strlen(name);
  str = strdup(str);
  for (found = strstr(str, "\\N"); found != NULL; found = strstr(str, "\\N")) {
    /* Ensure we have enough memory to do the replacement. */
    size_t min_len = strlen(str) + namelen - 2 + 1;
    if (allocated < min_len) {
      allocated = min_len;
      result = (char *) realloc(result, allocated);
    }

    /* Perform the replacement. */
    strncpy(result, str, found - str);
    resultp = &result[found - str];
    *resultp = '\0';
    strcat(resultp, name);
    resultp += namelen;
    strcpy(resultp, found + 2);

    /* Prepare for the next iteration. */
    free(str);
    str = strdup(result);
  }
  free(str);
  return result;
}

/* Draw a node, appropriately mapping its attributes. */
void draw_node (Agnode_t *node)
{
  int node_id = UBID(node);
  char *value;
  double size = 0.0;

  /* Create a new node. */
  ubigraph_new_vertex_w_id(node_id);

  /* Set the dot color scheme. */
  if (GET_NODE_VALUE("colorscheme"))
    setColorScheme(value);

  /* Handle fonts and labels. */
  value = agget(node, "label");
  ubigraph_set_vertex_attribute(node_id, "label", value == NULL ? agnameof(node) : replace_N(value, agnameof(node)));
  if (GET_NODE_VALUE("fontcolor"))
    ubigraph_set_vertex_attribute(node_id, "fontcolor", convert_color(value));
  if (GET_NODE_VALUE("fontname"))
    ubigraph_set_vertex_attribute(node_id, "fontfamily", value);
  if (GET_NODE_VALUE("fontsize"))
    ubigraph_set_vertex_attribute(node_id, "fontsize", value);

  /* Set the color only if the "style" attribute includes "filled". */
  if (GET_NODE_VALUE("style")) {
    if (strstr(value, "filled") != NULL) {
      char *value;

      if (GET_NODE_VALUE("fillcolor"))
        ubigraph_set_vertex_attribute(node_id, "color", convert_color(value));
    }
    if (strstr(value, "invis") != NULL)
      ubigraph_set_vertex_attribute(node_id, "visible", "false");
  }

  /* Map 2-D shapes to 3-D shapes. */
  if (GET_NODE_VALUE("shape"))
    ubigraph_set_vertex_attribute(node_id, "shape", convert_shape(value));

  /* Adjust the shape size. */
  if (GET_NODE_VALUE("width")) {
    double width = atof(value);
    if (width > size)
      size = width;
  }
  if (GET_NODE_VALUE("height")) {
    double height = atof(value);
    if (height > size)
      size = height;
  }
  if (size > 0) {
    char size_str[50];

    sprintf(size_str, "%31.5f", size);
    ubigraph_set_vertex_attribute(node_id, "size", size_str);
  }

  /* Honor URLs. */
  if (GET_NODE_VALUE("URL") || GET_NODE_VALUE("href"))
    ubigraph_set_vertex_attribute(node_id, "callback_left_doubleclick", value);
}

/* Draw an edge, appropriately mapping its attributes. */
void draw_edge (Agedge_t *edge)
{
  int edge_id = UBID(edge);
  char *value;

  /* Create a new edge. */
  ubigraph_new_edge_w_id(edge_id, UBID(agtail(edge)), UBID(aghead(edge)));

  /* Set the dot color scheme. */
  if (GET_EDGE_VALUE("colorscheme"))
    setColorScheme(value);

  /* Set the edge color. */
  if (GET_EDGE_VALUE("color")) {
    if (!strcmp(value, "invis"))
      ubigraph_set_edge_attribute(edge_id, "visible", "false");
    else
      ubigraph_set_edge_attribute(edge_id, "color", convert_color(value));
  }

  /* Handle fonts and labels. */
  if (GET_EDGE_VALUE("label")) {
    if (GET_EDGE_VALUE("label"))
      ubigraph_set_edge_attribute(edge_id, "label", value);
    if (GET_EDGE_VALUE("fontcolor"))
      ubigraph_set_edge_attribute(edge_id, "fontcolor", convert_color(value));
    if (GET_EDGE_VALUE("fontname"))
      ubigraph_set_edge_attribute(edge_id, "fontfamily", value);
    if (GET_EDGE_VALUE("fontsize"))
      ubigraph_set_edge_attribute(edge_id, "fontsize", value);
  }

  /* Set the edge width. */
  if (GET_EDGE_VALUE("penwidth"))
    ubigraph_set_edge_attribute(edge_id, "width", value);

  /* Set the edge style. */
  if (GET_EDGE_VALUE("style")) {
    if (strstr(value, "solid") != NULL)
      ubigraph_set_edge_attribute(edge_id, "stroke", "solid");
    else if (strstr(value, "dashed") != NULL)
      ubigraph_set_edge_attribute(edge_id, "stroke", "dashed");
    else if (strstr(value, "dotted") != NULL)
      ubigraph_set_edge_attribute(edge_id, "stroke", "dotted");
    else if (strstr(value, "invis") != NULL)
      ubigraph_set_edge_attribute(edge_id, "stroke", "none");
  }

  /* Set the edge strength. */
  if (GET_EDGE_VALUE("weight"))
    ubigraph_set_edge_attribute(edge_id, "strength", value);

  /* Set the arrowhead size and position. */
  if (GET_EDGE_VALUE("arrowsize")) {
    ubigraph_set_edge_attribute(edge_id, "arrow_radius", value);
    ubigraph_set_edge_attribute(edge_id, "arrow_length", value);
  }
  if (GET_EDGE_VALUE("dir")) {
    /* dir=forward is the default, and Ubigraph doesn't support dir=both. */
    if (!strcmp(value, "back"))
      ubigraph_set_edge_attribute(edge_id, "arrow_reverse", "true");
    else if (!strcmp(value, "none"))
      ubigraph_set_edge_attribute(edge_id, "arrow", "false");
  }
}

int main (int argc, const char *argv[])
{
  Agraph_t *graph;
  Agnode_t *node;

  /* Parse the specified dot file. */
  progname = argv[0];
  graph = read_dot(argc < 2 ? "-" : argv[1]);

  /* Establish a connection to the Ubigraph server and set default options. */
  initialize_ubigraph(graph);

  /* Draw all of the nodes in the entire graph. */
  for (node = agfstnode(graph); node != NULL; node = agnxtnode(graph, node))
    draw_node(node);

  /* Draw all of the edges in the entire graph. */
  for (node = agfstnode(graph); node != NULL; node = agnxtnode(graph, node)) {
    Agedge_t *edge;
    for (edge = agfstout(graph, node); edge != NULL; edge = agnxtout(graph, edge))
      draw_edge(edge);
  }

  /* Exit cleanly. */
  agclose(graph);
  return 0;
}
